File: snow_pkg/src/openalea/SnowDensity.py
# coding: utf8
from copy import copy

import numpy
from math import *

def model_SnowDensity(ps_t1 = 0.0,
         Sdepth_t1 = 0.0,
         Sdry_t1 = 0.0,
         Swet_t1 = 0.0):
    """
     - Name: SnowDensity -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Density of snow cover calculation
                 * Author: STICS
                 * Reference: -
                 * Institution: INRA
                 * Abstract: -
     - inputs:
                 * name: ps_t1
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : kg/m**3
                               ** description : density of snow cover in previous day
                 * name: Sdepth_t1
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 5000.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : m
                               ** description : snow cover depth Calculation in previous day
                 * name: Sdry_t1
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 500.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : mmW
                               ** description : water in solid state in the snow cover in previous day
                 * name: Swet_t1
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : mmW
                               ** description : water in liquid state in the snow cover
     - outputs:
                 * name: ps
                               ** min : 0.0
                               ** variablecategory : state
                               ** max : 500.0
                               ** uri : 
                               ** datatype : DOUBLE
                               ** unit : kg/m**3
                               ** description : density of snow cover
    """

    ps = 0.0
    if abs(Sdepth_t1) > 0.0:
        if abs(Sdry_t1 + Swet_t1) > 0.0:
            ps = (Sdry_t1 + Swet_t1) / Sdepth_t1
        else:
            ps = ps_t1
    return ps

File: snow_pkg/src/f90/SnowDensity.f90
MODULE Snowdensitymod
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_SnowDensity(ps_t1, &
        Sdepth_t1, &
        Sdry_t1, &
        Swet_t1, &
        ps)
        IMPLICIT NONE
        REAL, INTENT(IN) :: ps_t1
        REAL, INTENT(IN) :: Sdepth_t1
        REAL, INTENT(IN) :: Sdry_t1
        REAL, INTENT(IN) :: Swet_t1
        REAL, INTENT(OUT) :: ps
        !- Description:
    !            * Title: Density of snow cover calculation
    !            * Author: STICS
    !            * Reference: -
    !            * Institution: INRA
    !            * Abstract: -
        !- inputs:
    !            * name: ps_t1
    !                          ** description : density of snow cover in previous day
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 100.0
    !                          ** unit : kg/m**3
    !                          ** uri : 
    !            * name: Sdepth_t1
    !                          ** description : snow cover depth Calculation in previous day
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : m
    !                          ** uri : 
    !            * name: Sdry_t1
    !                          ** description : water in solid state in the snow cover in previous day
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 500.0
    !                          ** unit : mmW
    !                          ** uri : 
    !            * name: Swet_t1
    !                          ** description : water in liquid state in the snow cover
    !                          ** inputtype : variable
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 100.0
    !                          ** unit : mmW
    !                          ** uri : 
        !- outputs:
    !            * name: ps
    !                          ** description : density of snow cover
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 500.0
    !                          ** unit : kg/m**3
    !                          ** uri : 
        ps = 0.0
        IF(ABS(Sdepth_t1) .GT. 0.0) THEN
            IF(ABS(Sdry_t1 + Swet_t1) .GT. 0.0) THEN
                ps = (Sdry_t1 + Swet_t1) / Sdepth_t1
            ELSE
                ps = ps_t1
            END IF
        END IF
    END SUBROUTINE model_SnowDensity

END MODULE
